### PR TITLE
Dev to master

### PR DESCRIPTION
**Pull request by 3ch023**

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/InstanceIdProvider.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import org.apache.sling.api.SlingHttpServletRequest;
+
+/**
+ * Component demandware instance id service to identify the instance id of configured demandware.
+ */
+public interface InstanceIdProvider {
+    
+    String getInstanceId(Page page);
+    
+    String getInstanceId(SlingHttpServletRequest request);
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/InstanceIdProvider.java:30]: @allwardt 

the instanceID should not be derived from the referrer. If we need the ID for subsequent requests... maybe it can be transmitted explicitly via selector or request parameter? what is the use case?

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientImpl.java [hunk 2]
```diff
@@ -79,7 +84,7 @@
     @Property(label = "Local network interface to be used")
     private static final String PROTOCOL_INTERFACE = "interface";
 
-    @Property(label = "SSL version", value = "TLSv1.2")
+    @Property(label = "SSL version", value = "TLSv1.1")
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientImpl.java:85]: Why are we changing back to 1.1 ?

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientImpl.java [hunk 3]
```diff
@@ -236,9 +249,21 @@ public CloseableHttpClient getHttpClient() {
...
+        return assetDownloadEndpoint;
+    }
+
     @Activate
     protected void activate(Map<String, Object> configuration) {
         instanceEndPoint = PropertiesUtil.toString(configuration.get(INSTANCE_ENDPOINT), null);
+        assetDownloadEndpoint = PropertiesUtil.toString(configuration.get(ASSET_DOWNLOAD_ENDPOINT),"");
+        instanceId = PropertiesUtil.toString(configuration.get(INSTANCE_ID), null);
         protocolInterface = PropertiesUtil.toString(configuration.get(PROTOCOL_INTERFACE), null);
         protocolSSL = StringUtils.trimToNull(PropertiesUtil.toString(configuration.get(PROTOCOL_SSL), "TLSv1.2"));
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientImpl.java]: what is the default? "1.2" or "1.1" as defined in the property?

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientImpl.java [hunk 4]
```diff
@@ -93,6 +98,12 @@
 
     @Property(label = "Key password", description = "Leave empty for no password")
     private static final String KEY_PWD = "key.password";
+    
+    @Property(label = "Asset download endpoint")
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientImpl.java]: can we provide a default or a hint what to enter here?

File: content/src/main/content/jcr_root/apps/commerce/demandware/config.author/com.adobe.cq.commerce.demandware.connection.DemandwareClientImpl-defaultInstance.config [hunk 5]
```diff
@@ -0,0 +1,3 @@
+endpoint="{FILL ME}.demandware.net"
+ssl="TLSv1.1"
+instance.id="defaultInstance"
```
Comment (ackoch) [inline content/src/main/content/jcr_root/apps/commerce/demandware/config.author/com.adobe.cq.commerce.demandware.connection.DemandwareClientImpl-defaultInstance.config]: please add all properties the service provides to this demo-file

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/PreviewServiceImpl.java [hunk 6]
```diff
@@ -260,7 +261,7 @@ protected CredentialsProvider createCredentialsProvider(Map<String, Object> conf
         if (PropertiesUtil.toString(configuration.get(STORFRONT_PROTECTION_USER), null) != null && PropertiesUtil
                 .toString(configuration.get(STORFRONT_PROTECTION_PASSWORD), null) != null) {
             final CredentialsProvider credsProvider = new BasicCredentialsProvider();
-            credsProvider.setCredentials(new AuthScope(demandwareClient.getEndpoint(), AuthScope.ANY_PORT),
+            credsProvider.setCredentials(new AuthScope(clientProvider.getDefaultClient().getEndpoint(), AuthScope.ANY_PORT),
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/PreviewServiceImpl.java]: we have probably two instances here.. and maybe not even a default one. this would mean, we cannot preview on a multi-instance setup.

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/PreviewServiceImpl.java [hunk 7]
```diff
@@ -225,7 +226,7 @@ private String getPreviewContent(final String endPoint, final Resource resource,
     }
 
     private String getInstanceEndPoint() {
-        return demandwareClient.getEndpoint();
+        return clientProvider.getDefaultClient().getEndpoint();
     }
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/PreviewServiceImpl.java:205]: we have probably two instances here.. and maybe not even a default one. this would mean, we cannot preview on a multi-instance setup.

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/InstanceIdProviderImpl.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    }
+    
+    @Override
+    public String getInstanceId(final AgentConfig config) {
+        return Optional.ofNullable(config.getTransportURI())
+                .map(uri -> uri.replace(DWRE_SCHEME, StringUtils.EMPTY))
+                .orElse(StringUtils.EMPTY);
+    }
+    
+    private String getReferrerPath(final SlingHttpServletRequest request) {
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/InstanceIdProviderImpl.java:71]: currently not used... and not a reliable technique

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	@Override
+	public Optional<DemandwareClient> getClientForSpecificInstance(final String instanceId) {
+		DemandwareClient demandwareClient = demandwareClients.get(instanceId.replace("/", ""));
+		if (demandwareClient == null) {
+			LOG.error("DemandwareClient not found for instanceId [{}]", instanceId);
+			return Optional.empty();
+		}
+		return Optional.of(demandwareClient);
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java]: this is not "optional" when you can't provide a client, you cannot continue in the upper processes. we should throw an exception. not having a service is a serious problem. btw: when you throw excpetions, don't log.

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+			return clientOpt.get();
+		}
+
+		LOG.error("Failed to get DemandwareClient - no configuration found.");
+		return null;
+	}
+
+	@Override
+	public Optional<DemandwareClient> getClientForSpecificInstance(final String instanceId) {
+		DemandwareClient demandwareClient = demandwareClients.get(instanceId.replace("/", ""));
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java]: we are defining the names ourselves. where does the "/" come from? and why do we have to remove it?

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+			cardinality = ReferenceCardinality.MANDATORY_MULTIPLE, // TODO check if works with one
+			policy = ReferencePolicy.DYNAMIC)
+	protected HashMap<String, DemandwareClient> demandwareClients;
+
+	/**
+	 * Returns DemandwareClient service, configured with "default" instance id
+	 * or, if not present, first found DemandwareClient service.
+	 *
+	 * @return
+	 *///TODO remove usages
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java]: remove TDODs

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+@Component(label = "Demandware Demandware Client Provider", immediate = true)
+@Service(value = DemandwareClientProvider.class)
+public class DemandwareClientProviderImpl implements DemandwareClientProvider {
+	private static final Logger LOG = LoggerFactory.getLogger(DemandwareClientProviderImpl.class);
+	private static final String DWRE_SCHEME = "demandware://";
+	public static final String INSTANCE_ID_DEFAULT = "default";
+	@Reference(referenceInterface = DemandwareClient.class,
+			bind = "bindDemandwareClient",
+			unbind = "unbindDemandwareClient",
+			cardinality = ReferenceCardinality.MANDATORY_MULTIPLE, // TODO check if works with one
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java]: remove TODOs

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/StaticContentProxyServlet.java [hunk 13]
```diff
@@ -52,18 +52,18 @@
...
+    private DemandwareClientProvider clientProvider;
 
     @Override
     protected void doGet(SlingHttpServletRequest request, SlingHttpServletResponse response) throws ServletException,
             IOException {
 
         RequestPathInfo pathInfo = request.getRequestPathInfo();
         LOG.debug("Proxy static content for {}", pathInfo.toString());
-        final String remoteUri = DemandwareClient.DEFAULT_SCHEMA + demandwareClient.getEndpoint() + pathInfo.getResourcePath() +
+        final String remoteUri = DemandwareClient.DEFAULT_SCHEMA + clientProvider.getDefaultClient().getEndpoint() + pathInfo.getResourcePath() +
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/preview/StaticContentProxyServlet.java]: we have probably two instances here.. and maybe not even a default one. this would mean, we cannot preview on a multi-instance setup.

File: bundles/init/pom.xml [hunk 14]
```diff
@@ -47,6 +47,13 @@
 
     <build>
         <plugins>
+            <plugin>
+                <groupId>org.apache.sling</groupId>
+                <artifactId>maven-sling-plugin</artifactId>
+                <configuration>
+                    <slingUrl>http://localhost:4502/system/console/install</slingUrl>
```
Comment (ackoch) [inline bundles/init/pom.xml]: make configurable

File: bundles/init/pom.xml [hunk 15]
```diff
@@ -47,6 +47,13 @@
 
     <build>
         <plugins>
+            <plugin>
```
Comment (ackoch) [inline bundles/init/pom.xml]: bind to profile

File: bundles/init/src/main/java/com/adobe/cq/commerce/demandware/init/InitServlet.java [hunk 16]
```diff
@@ -59,26 +60,28 @@
...
     @Override
-    protected void doPost(SlingHttpServletRequest request, SlingHttpServletResponse response) throws
-        ServletException, IOException {
+    protected void doPost(SlingHttpServletRequest request, SlingHttpServletResponse response) throws IOException {
         LOG.debug("Init Demandware Sandbox");
         ValueMap config = request.getResource().getValueMap();
-
-        // get assets and push to webdav
+	    final String instanceId = config.get(PROPERTY_DW_INSTANCE_ID, String.class);
+	    final Optional<DemandwareClient> dwClient = clientProvider.getClientForSpecificInstance(instanceId);
```
Comment (ackoch) [inline bundles/init/src/main/java/com/adobe/cq/commerce/demandware/init/InitServlet.java]: "Optional" is not useful here - especially as we test that you have one later on..

File: bundles/init/src/main/java/com/adobe/cq/commerce/demandware/init/InitServlet.java [hunk 17]
```diff
@@ -59,26 +60,28 @@
...
+	    // get assets and push to webdav
         final String webDAVEndpoint = config.get("assetWebDAV", String.class);
         final String[] assetURIs = config.get("assetUris", String[].class);
-        if (StringUtils.isNotEmpty(webDAVEndpoint) && ArrayUtils.isNotEmpty(assetURIs)) {
-            for (String assetURI : assetURIs) {
-                upload(request, webDAVEndpoint, assetURI);
+        if (StringUtils.isNotEmpty(webDAVEndpoint) && assetURIs != null && ArrayUtils.isNotEmpty(assetURIs)
+		        && dwClient.isPresent()) {
+	        for (String assetURI : assetURIs) {
+                upload(request, webDAVEndpoint, assetURI, dwClient.get());
```
Comment (ackoch) [inline bundles/init/src/main/java/com/adobe/cq/commerce/demandware/init/InitServlet.java]: we shouldn't resolve on each request. better not to have an optional

File: bundles/init/src/main/java/com/adobe/cq/commerce/demandware/init/InitServlet.java [hunk 18]
```diff
@@ -101,20 +104,21 @@ protected void doPost(SlingHttpServletRequest request, SlingHttpServletResponse
...
-        final HttpClient httpClient = demandwareClient.getHttpClientBuilder().build();
+        final HttpClient httpClient = clientProvider.getDefaultClient().getHttpClientBuilder().build();
         try {
             // get the from AEM content
             final HttpResponse response = httpClient.execute(localHost, requestBuilder.build());
 
             if (response != null && response.getStatusLine().getStatusCode() == HttpStatus.SC_OK) {
                 final HttpEntity entity = response.getEntity();
-                HttpClientBuilder httpClientBuilder = demandwareClient.getHttpClientBuilder();
+                HttpClientBuilder httpClientBuilder = dwClient.getHttpClientBuilder();
```
Comment (ackoch) [inline bundles/init/src/main/java/com/adobe/cq/commerce/demandware/init/InitServlet.java:119]: why are we using "dwClient" here and the "defaultClient" above?

File: bundles/init/src/main/java/com/adobe/cq/commerce/demandware/init/InitServlet.java [hunk 19]
```diff
@@ -101,20 +104,21 @@ protected void doPost(SlingHttpServletRequest request, SlingHttpServletResponse
...
-    private void upload(SlingHttpServletRequest slingRequest, String endPoint, String assetURI) {
+    private void upload(SlingHttpServletRequest slingRequest, String endPoint, String assetURI,
+                        DemandwareClient dwClient) {
         final RequestBuilder requestBuilder = RequestBuilder.get();
         requestBuilder.setUri(assetURI);
         final Cookie token = slingRequest.getCookie("login-token");
         requestBuilder.addHeader(new BasicHeader("Cookie", token.getName() + "=" + token.getValue()));
         final HttpHost localHost = new HttpHost("localhost", 4502);
-        final HttpClient httpClient = demandwareClient.getHttpClientBuilder().build();
+        final HttpClient httpClient = clientProvider.getDefaultClient().getHttpClientBuilder().build();
```
Comment (ackoch) [inline bundles/init/src/main/java/com/adobe/cq/commerce/demandware/init/InitServlet.java]: we can use dwClient here... If we don't have a client, this method wouldn't have been called in the first place

File: bundles/libs/pom.xml [hunk 20]
```diff
@@ -43,6 +43,13 @@
     <!-- ====================================================================== -->
     <build>
         <plugins>
+            <plugin>
```
Comment (ackoch) [inline bundles/libs/pom.xml]: bind to profile

File: bundles/libs/pom.xml [hunk 21]
```diff
@@ -43,6 +43,13 @@
...
+                <groupId>org.apache.sling</groupId>
+                <artifactId>maven-sling-plugin</artifactId>
+                <configuration>
+                    <slingUrl>http://localhost:4502/system/console/install</slingUrl>
```
Comment (ackoch) [inline bundles/libs/pom.xml]: make configurable via settings.xml

File: bundles/pim/pom.xml [hunk 22]
```diff
@@ -42,6 +42,13 @@
     <!-- ====================================================================== -->
     <build>
         <plugins>
+            <plugin>
```
Comment (ackoch) [inline bundles/pim/pom.xml:47]: profile

File: bundles/pim/pom.xml [hunk 23]
```diff
@@ -42,6 +42,13 @@
...
+                <groupId>org.apache.sling</groupId>
+                <artifactId>maven-sling-plugin</artifactId>
+                <configuration>
+                    <slingUrl>http://localhost:4502/system/console/install</slingUrl>
```
Comment (ackoch) [inline bundles/pim/pom.xml]: settings.xml

File: bundles/pim/src/main/java/com/adobe/cq/commerce/demandware/pim/impl/DemandwareImageAssetHandler.java [hunk 24]
```diff
@@ -16,16 +16,16 @@
 
 package com.adobe.cq.commerce.demandware.pim.impl;
```
Comment (ackoch) [inline bundles/pim/src/main/java/com/adobe/cq/commerce/demandware/pim/impl/DemandwareImageAssetHandler.java:17]: i need a live demo of the code :-/

File: bundles/replication/pom.xml [hunk 25]
```diff
@@ -43,6 +43,13 @@
     <!-- ====================================================================== -->
     <build>
         <plugins>
+            <plugin>
```
Comment (ackoch) [inline bundles/replication/pom.xml]: profile

File: bundles/replication/pom.xml [hunk 26]
```diff
@@ -43,6 +43,13 @@
...
+                <groupId>org.apache.sling</groupId>
+                <artifactId>maven-sling-plugin</artifactId>
+                <configuration>
+                    <slingUrl>http://localhost:4502/system/console/install</slingUrl>
```
Comment (ackoch) [inline bundles/replication/pom.xml]: settings.xml

File: bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/AbstractOCAPITransportPlugin.java [hunk 27]
```diff
@@ -94,22 +93,17 @@
...
-    protected DemandwareClient demandwareClient;
+    private InstanceIdProvider instanceIdProvider;
 
     private Map<String, Comparable<Object>> accessTokenProvidersProps =
             new ConcurrentSkipListMap<>(Collections.reverseOrder());
     private Map<Comparable<Object>, AccessTokenProvider> accessTokenProviders =
             new ConcurrentSkipListMap<>(Collections.reverseOrder());
 
-    private String accessTokenProviderId;
+    private String atpClientId;
```
Comment (ackoch) [inline bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/AbstractOCAPITransportPlugin.java]: do not use abbreviations!

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+	protected HashMap<String, DemandwareClient> demandwareClients;
+
+	/**
+	 * Returns DemandwareClient service, configured with "default" instance id
+	 * or, if not present, first found DemandwareClient service.
+	 *
+	 * @return
+	 *///TODO remove usages
+	public DemandwareClient getDefaultClient() {
+		Optional<DemandwareClient> client = getClientForSpecificInstance(INSTANCE_ID_DEFAULT);
```
Comment (ackoch) [inline bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java]: that's not the right usage of "Optional"

File: bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/AbstractOCAPITransportPlugin.java [hunk 29]
```diff
@@ -205,18 +200,22 @@ public boolean deliver(JSONObject delivery, AgentConfig config, ReplicationLog l
...
             throw new ReplicationException("No request method provided");
         }
         try {
             // construct the OCAPI request
             final StringBuilder transportUriBuilder = new StringBuilder();
-            transportUriBuilder.append(DemandwareClient.DEFAULT_SCHEMA).append(demandwareClient.getEndpoint());
+            final String endpoint = clientProvider.getClientForSpecificInstance(dwInstanceId)
+                    .map(DemandwareClient::getEndpoint)
+                    .orElse(StringUtils.EMPTY);
+            transportUriBuilder.append(DemandwareClient.DEFAULT_SCHEMA).append(endpoint);
```
Comment (ackoch) [inline bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/AbstractOCAPITransportPlugin.java:217]: is an empty endpoint really a valid option?

File: bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/ContentAssetFolderPlugin.java [hunk 30]
```diff
@@ -76,8 +80,10 @@ public boolean deliver(JSONObject delivery, AgentConfig config, ReplicationLog l
                         log.info("Assign content asset %s to folder %s (%s)", id, folder, action.getType().getName());
 
                         // construct URL for folder assignment
-                        final String transportUriBuilder = DemandwareClient.DEFAULT_SCHEMA + demandwareClient.getEndpoint() + getOCApiPath() + getOCApiVersion() + constructEndpointURL(
-                                delivery.getString(DemandwareCommerceConstants.ATTR_API_ENDPOINT), folder, delivery);
+                        final String endpoint = clientProvider.getClientForSpecificInstance(instanceId.getInstanceId(config))
+                                .map(DemandwareClient::getEndpoint).orElse(StringUtils.EMPTY);
```
Comment (ackoch) [inline bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/ContentAssetFolderPlugin.java]: we need an endpoint. this is not optional. it must not be empty.

File: bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/WebDAVTransportPlugin.java [hunk 31]
```diff
@@ -76,9 +74,6 @@
     @Property(label = "WebDAV user password")
     private static final String WEBDAV_PASSWORD = "webdav.password";
 
-    @Reference
-    DemandwareClient demandwareClient;
-
     private String webDavEndpoint;
     private String webDavUser;
```
Comment (ackoch) [inline bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/WebDAVTransportPlugin.java]: we have to make that a factory when it has local variables. or we move the credentials to the demandware-client

File: bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/WebDAVTransportPlugin.java [hunk 32]
```diff
@@ -93,20 +88,18 @@ String getContentType() {
...
-        return demandwareClient;
-    }
-
     @Override
     public boolean deliver(JSONObject delivery, AgentConfig config, ReplicationLog log, ReplicationAction action)
         throws ReplicationException {
 
         // construct the WebDAV request
         String path = null;
+        final String endpoint = StringUtils.isNotEmpty(webDavEndpoint)
```
Comment (ackoch) [inline bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/WebDAVTransportPlugin.java]: this is not a good fallback strategy

File: bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/WebDAVTransportPlugin.java [hunk 33]
```diff
@@ -245,6 +238,15 @@ private void deleteWevDAV(HttpClientBuilder httpClientBuilder, String transportU
         }
     }
 
+    final String getDemandwareClientEndpoint(final AgentConfig config) {
+        final Optional<DemandwareClient> demandwareClient = clientProvider.getClientForSpecificInstance(config);
```
Comment (ackoch) [inline bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/WebDAVTransportPlugin.java]: this is not optional

File: bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/oauth/AccessTokenProviderImpl.java [hunk 34]
```diff
@@ -207,8 +218,12 @@ protected void activate(Map<String, Object> props)
 
         if (HTTPS.equals(uri.getScheme())) {
             relativeUri = uri.toString();
-
-            HttpClientBuilder builder = demandwareClient.getHttpClientBuilder();
+            Optional<DemandwareClient> dwClient = clientProvider.getClientForSpecificInstance(instanceId);
```
Comment (ackoch) [inline bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/oauth/AccessTokenProviderImpl.java]: This is not optional

File: content-multi-sample/src/main/content/jcr_root/apps/commerce/demandware/config.author/com.adobe.cq.commerce.demandware.replication.transport.WebDAVTransportPlugin.config [hunk 35]
```diff
@@ -0,0 +1,2 @@
+webdav.user="{FILL ME}"
```
Comment (ackoch) [inline content-multi-sample/src/main/content/jcr_root/apps/commerce/demandware/config.author/com.adobe.cq.commerce.demandware.replication.transport.WebDAVTransportPlugin.config]: this must be either configured in the demandwareclient or the webdav plugin needs to be a factory: the two instances might have to use different credentials..

File: content-multi-sample/src/main/content/jcr_root/content/eu-sitegenesis/_rep_policy.xml [hunk 36]
```diff
@@ -0,0 +1,23 @@
...
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~-->
+<jcr:root xmlns:jcr="http://www.jcp.org/jcr/1.0" xmlns:rep="internal"
+          jcr:primaryType="rep:ACL">
+    <allowEUSitegenesis
+            jcr:primaryType="rep:GrantACE"
+            rep:principalName="dwre-replication-agent-eu-sitegenesis"
+            rep:privileges="{Name}[jcr:read,crx:replicate]"/>
```
Comment (ackoch) [inline content-multi-sample/src/main/content/jcr_root/content/eu-sitegenesis/_rep_policy.xml]: afaik, "read" is enough, no need to "replicate"

File: content-sample/src/main/content/jcr_root/content/sitegenesis/_rep_policy.xml [hunk 37]
```diff
@@ -0,0 +1,23 @@
...
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~-->
+<jcr:root xmlns:jcr="http://www.jcp.org/jcr/1.0" xmlns:rep="internal"
+          jcr:primaryType="rep:ACL">
+    <allowSitegenesisAgent
+            jcr:primaryType="rep:GrantACE"
+            rep:principalName="dwre-replication-agent-sitegenesis"
+            rep:privileges="{Name}[jcr:read,crx:replicate]"/>
```
Comment (ackoch) [inline content-sample/src/main/content/jcr_root/content/sitegenesis/_rep_policy.xml]: "read" should suffice

File: content-sample/src/main/content/jcr_root/etc/demandware/.content.xml [hunk 38]
```diff
@@ -21,9 +21,10 @@
                 jcr:primaryType="nt:unstructured"
```
Comment (ackoch) [inline content-sample/src/main/content/jcr_root/etc/demandware/.content.xml:21]: what's that file for?

File: content/src/main/content/jcr_root/apps/commerce/demandware/config.author/com.adobe.cq.commerce.demandware.oauth.accesstoken.provider-default.config [hunk 39]
```diff
@@ -0,0 +1,10 @@
+auth.access.token.request="https://%s/dw/oauth2/access_token"
```
Comment (ackoch) [inline content/src/main/content/jcr_root/apps/commerce/demandware/config.author/com.adobe.cq.commerce.demandware.oauth.accesstoken.provider-default.config]: move to separate "config" module

File: content/src/main/content/jcr_root/content/_rep_policy.xml [hunk 40]
```diff
@@ -16,7 +16,7 @@
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~-->
 <jcr:root xmlns:jcr="http://www.jcp.org/jcr/1.0" xmlns:rep="internal"
           jcr:primaryType="rep:ACL">
-    <allow39
+    <allowDemandwareService
```
Comment (ackoch) [inline content/src/main/content/jcr_root/content/_rep_policy.xml]: why is that necessary?

File: content/src/main/content/jcr_root/etc/demandware/.content.xml [hunk 41]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ackoch) [inline content/src/main/content/jcr_root/etc/demandware/.content.xml]: why do we need this?

File: content/src/main/content/jcr_root/etc/replication/agents.author/demandware/.content.xml [hunk 42]
```diff
@@ -22,11 +22,11 @@
             jcr:primaryType="nt:unstructured"
```
Comment (ackoch) [inline content/src/main/content/jcr_root/etc/replication/agents.author/demandware/.content.xml:22]: move to "config" module

File: content/src/main/content/jcr_root/home/users/.content.xml [hunk 43]
```diff
@@ -16,4 +16,12 @@
...
 <jcr:root xmlns:jcr="http://www.jcp.org/jcr/1.0" xmlns:rep="internal"
     jcr:mixinTypes="[rep:AccessControllable]"
-    jcr:primaryType="rep:AuthorizableFolder"/>
+          jcr:primaryType="rep:AuthorizableFolder">
+    <rep:policy
+            jcr:primaryType="rep:ACL">
+        <allowDWREReplicationService
+                jcr:primaryType="rep:GrantACE"
+                rep:principalName="dwre-replication-service"
+                rep:privileges="{Name}[jcr:all]"/>
```
Comment (ackoch) [inline content/src/main/content/jcr_root/home/users/.content.xml]: this gives the replication system-account full admin rights on the users! why?

File: parent/pom.xml [hunk 44]
```diff
@@ -535,7 +535,7 @@
             <dependency>
                 <groupId>com.adobe.aem</groupId>
                 <artifactId>uber-jar</artifactId>
-                <version>6.4.1</version>
+                <version>6.4.3</version>
```
Comment (ackoch) [inline parent/pom.xml]: we can make that 6.4.5 already

**Review by ackoch — CHANGES_REQUESTED** [review]
- minor bugs and general cleanup required
- preview seems to work only with one instance 
- use of "Optional" not ideal

File: bundles/init/src/main/java/com/adobe/cq/commerce/demandware/init/InitServlet.java (hunk 18 above)
Comment (3ch023) [reply bundles/init/src/main/java/com/adobe/cq/commerce/demandware/init/InitServlet.java:119]: hi, I fixed this in dev-with-preview branch, as soon as we merge it to dev it should be ok.

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java (hunk 28 above)
Comment (3ch023) [reply bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java]: getDefaultClient() was temporary methond, it is removed completely now

File: bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/AbstractOCAPITransportPlugin.java (hunk 29 above)
Comment (3ch023) [reply bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/AbstractOCAPITransportPlugin.java:217]: Will change it to `throw new ReplicationException` in the client not found

File: bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/ContentAssetFolderPlugin.java (hunk 30 above)
Comment (3ch023) [reply bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/ContentAssetFolderPlugin.java]: We can throw ReplicationException if Demandware Client not found. I didn't request this before because it was not thrown in the initial implementation and we tried to change as little as possible.

Additionally, getting the Demandware Client can be moved out of the for loop.

File: content-multi-sample/src/main/content/jcr_root/content/eu-sitegenesis/_rep_policy.xml (hunk 36 above)
Comment (3ch023) [reply content-multi-sample/src/main/content/jcr_root/content/eu-sitegenesis/_rep_policy.xml]: ok, I will try with only jcr:read

File: content-sample/src/main/content/jcr_root/etc/demandware/.content.xml (hunk 38 above)
Comment (3ch023) [reply content-sample/src/main/content/jcr_root/etc/demandware/.content.xml:21]: it creates a page with "Demandware Instance initializer"

File: content/src/main/content/jcr_root/etc/demandware/.content.xml (hunk 41 above)
Comment (3ch023) [reply content/src/main/content/jcr_root/etc/demandware/.content.xml]: Here is some info: https://github.com/adobe/commerce-salesforce/wiki/Set-up-&-configuration-instructions#initial-template-provisioning

File: content/src/main/content/jcr_root/home/users/.content.xml (hunk 43 above)
Comment (3ch023) [reply content/src/main/content/jcr_root/home/users/.content.xml]: my poor implementation. 'dwre-replication-service' used to create oauth subnode under instance-specific-user node, so I granted access. I will rewrite AccessTokenProviderImpl and impersonate in specific user.

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientImpl.java (hunk 3 above)
Comment (ackoch) [reply bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientImpl.java]: if(instanceEndpoint == null || asset..){
	
	throw new Exception("instanceEndpoint:" + " instanceEndpoint" + ... )
}

File: bundles/init/pom.xml [hunk 45]
```diff
@@ -66,9 +73,8 @@
         <dependency>
             <groupId>com.adobe.cq.commerce</groupId>
             <artifactId>cq-commerce-demandware-core</artifactId>
-            <version>1.3.7-SNAPSHOT</version>
+            <version>2.0.0-SNAPSHOT</version>
```
Comment (ackoch) [inline bundles/init/pom.xml:89]: Please consistently use 2.0 everywhere

File: bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/WebDAVTransportPlugin.java (hunk 31 above)
Comment (ackoch) [reply bundles/replication/src/main/java/com/adobe/cq/commerce/demandware/replication/transport/WebDAVTransportPlugin.java]: agreed to put it into the client

File: content-sample/src/main/content/jcr_root/etc/demandware/.content.xml (hunk 38 above)
Comment (ackoch) [reply content-sample/src/main/content/jcr_root/etc/demandware/.content.xml:21]: move ti sample-content

File: bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/InstanceIdProviderImpl.java (hunk 8 above)
Comment (ackoch) [reply bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/InstanceIdProviderImpl.java:71]: @3ch023 : Please do not return an empty instance ID - that is not a valid ID. If we can't find an ID that is an Exception - we cannot continue any useful operation  without that ID, can we?
